src/tribler/core/components/ipv8/eva_protocol.py

```diff
@@ -144,32 +144,66 @@ class TransferResult:
     data: bytes
     nonce: int
 
+    def __str__(self):
+        return f'TransferResult(peer={self.peer}, info: {self.info}, data hash: {hash(self.data)}, nonce={self.nonce})'
+
+class TransferWindow:
+    def __init__(self, start: int, size: int):
+        self.data: List[Optional[bytes]] = [None for _ in range(size)]
+
+        self.size = size
+        self.start = start
+        self.processed: int = 0
+        self.last_window: bool = False
+
+    def add(self, index: int, block: bytes):
+        if self.data[index] is not None:
+            return
+        self.data[index] = block
+        self.processed += 1
+
+    def finished(self) -> bool:
+        return self.processed == self.size
+
+    def flush(self, data_list: List[bytes]) -> int:
```

Comment:
**thought:** It may be a tiny bit more flexible to have a method that returns a list of consecutive blocks instead of modifying the `data_list` as a side effect. In that case, inside the method, you can write something like:

```
def consecutive_blocks(self):
    for block in self.data:
        if block is None:
            break
        yield block
```

And outside of the method, you can do `transfer.data_list.extend(window.consecutive_blocks())`.

In that case, when creating a `TransferWindow`, instead of calculating `start=self.window.start + added` we can do `start=len(transfer.data_list)`, so calculating `added` is unnecessary.

Reply:
Agree, your design is better. 
I partially dislike my implementation of `flush` because it modified `data_list` and makes it mutable (which we should avoid).

Reply:
I've implemented your suggestion.